core/src/main/java/com/oath/oak/Chunk.java

```diff
@@ -94,199 +93,186 @@
+     *            value was not unlinked yet.
+     *            {@code ctx.valueState} will indicate the reason.
+     * @param key the key to look up
+     */
+    void lookUp(ThreadContext ctx, K key) {
+        // binary search sorted part of key array to quickly find node to start search at
+        // it finds previous-to-key
+        int curr = binaryFind(ctx.tempKey, key);
+        curr = (curr == NONE_NEXT) ? entrySet.getHeadNextIndex() : entrySet.getNextEntryIndex(curr);
+
+        // iterate until end of list (or key is found)
+        while (curr != NONE_NEXT) {
+            // compare current item's key to searched key
+            boolean isValid = entrySet.keyLookUp(ctx, curr);
+            assert isValid;
+            int cmp = comparator.compareKeyAndSerializedKey(key, ctx.key.getAllocByteBuffer());
+            // if item's key is larger - we've exceeded our key
+            // it's not in chunk - no need to search further
+            if (cmp < 0) {
+                // Reset lookup key state to INVALID
+                ctx.invalidate();
+                return;
+            }
+            // if keys are equal - we've found the item
+            else if (cmp == 0) {
+                // Update the lookup value state
+                entrySet.valueLookUp(ctx);
+                return;
+            }
+            // otherwise- proceed to next item
+            curr = entrySet.getNextEntryIndex(curr);
+        }
 
-    // TODO: to change OakRKeyBuffer to OakRValueBuffer
-    boolean setValueReference(int ei, OakDetachedReadKeyBuffer valueRef) {
-        return entrySet.setValueOutputRBuff(ei, valueRef);
+        // Reset lookup key state to INVALID
+        ctx.invalidate();
     }
 
     /**
      * binary search for largest-entry smaller than 'key' in sorted part of key array.
      *
-     * @return the index of the entry from which to start a linear search -
+     * @param tempKey a reusable buffer object for internal temporary usage
+     * @param key     the key to look up
+     * @return        the index of the entry from which to start a linear search -
      * if key is found, its previous entry is returned!
      * In cases when search from the head is needed, meaning:
      * (1) the given key is less or equal than the smallest key in the chunk OR
      * (2) entries are unsorted so there is a need to start from the beginning of the linked list
      * NONE_NEXT is going to be returned
      */
-    private int binaryFind(K key) {
+    private int binaryFind(EntrySet.KeyBuffer tempKey, K key) {
         int sortedCount = this.sortedCount.get();
+        // if there are no sorted keys, return the head entry for a regular linear search
+        if (sortedCount == 0) {
+            return NONE_NEXT;
+        }
+
         int headIdx = entrySet.getHeadNextIndex();
-        // if there are no sorted keys, or the first item is already larger than key -
+        entrySet.readKey(tempKey, headIdx);
```

Comment:
My initial comment was: "Why to add here readKey if we aren't maybe going to continue". Then I understood that this is needed for comparison. So again, I think we need some wrap method for comparison, that will look like just comparison of keys providing maybe entry indexes of the keys and thread context. It should return the normal compare int. Internally this method should to keyLookUp/readKey/etc.

Reply:
I followed your suggestion and made an auxiliary function: `compareKeyAndEntryIndex()` method to `Chunk`.

As a result, I made a few other modifications:
- Moved `ValueState` definition to `EntrySet`.
- Add `getValueState()` method to `EntrySet` (auxiliary private function).
- Remove redundant methods `isValueLinkNeeded()` and `isDeleteValueFinishNeeded()`. 
  * Instead, check `ctx.valueState` directly.
- rename `keyLookUp(ThreadContext)` to `readKey(ThreadContext)` and `valueLookUp(ThreadContext)` to `readValue(ThreadContext)`.
- rename in the comments anywhere that mentioned "look-up context" to "entry context". 
  * remove all reference to the word lookup except the method `Chunk.lookup()`.